--- linux-2.6.32/drivers/leds/leds-s3c24xx.c	2010-05-04 03:29:11.000000000 -0300
+++ linux-2.6.32/drivers/leds/leds-s3c24xx.c	2010-07-02 14:24:21.000000000 -0300
@@ -89,12 +89,12 @@
	led->pdata = pdata;

	/* no point in having a pull-up if we are always driving */
-
+	s3c2410_gpio_pullup(pdata->gpio, 1);
+
	if (pdata->flags & S3C24XX_LEDF_TRISTATE) {
		s3c2410_gpio_setpin(pdata->gpio, 0);
		s3c2410_gpio_cfgpin(pdata->gpio, S3C2410_GPIO_INPUT);
	} else {
-		s3c2410_gpio_pullup(pdata->gpio, 0);
		/* backlight led needs to be turned on early on */
		s3c2410_gpio_setpin(pdata->gpio, led->cdev.brightness);
		s3c2410_gpio_cfgpin(pdata->gpio, S3C2410_GPIO_OUTPUT);
--- linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus0-gpe11_12_13.c	2010-05-04 03:28:43.000000000 -0300
+++ linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus0-gpe11_12_13.c	2010-07-02 14:28:31.000000000 -0300
@@ -24,13 +24,13 @@
		s3c2410_gpio_cfgpin(S3C2410_GPE(13), S3C2410_GPE13_SPICLK0);
		s3c2410_gpio_cfgpin(S3C2410_GPE(12), S3C2410_GPE12_SPIMOSI0);
		s3c2410_gpio_cfgpin(S3C2410_GPE(11), S3C2410_GPE11_SPIMISO0);
-		s3c2410_gpio_pullup(S3C2410_GPE(11), 0);
-		s3c2410_gpio_pullup(S3C2410_GPE(13), 0);
+		s3c2410_gpio_pullup(S3C2410_GPE(11), 1);
+		s3c2410_gpio_pullup(S3C2410_GPE(13), 1);
	} else {
		s3c2410_gpio_cfgpin(S3C2410_GPE(13), S3C2410_GPIO_INPUT);
		s3c2410_gpio_cfgpin(S3C2410_GPE(11), S3C2410_GPIO_INPUT);
-		s3c2410_gpio_pullup(S3C2410_GPE(11), 1);
-		s3c2410_gpio_pullup(S3C2410_GPE(12), 1);
-		s3c2410_gpio_pullup(S3C2410_GPE(13), 1);
+		s3c2410_gpio_pullup(S3C2410_GPE(11), 0);
+		s3c2410_gpio_pullup(S3C2410_GPE(12), 0);
+		s3c2410_gpio_pullup(S3C2410_GPE(13), 0);
	}
 }
--- linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus1-gpd8_9_10.c	2010-05-04 03:28:43.000000000 -0300
+++ linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus1-gpd8_9_10.c	2010-07-02 14:28:25.000000000 -0300
@@ -26,13 +26,13 @@
		s3c2410_gpio_cfgpin(S3C2410_GPD(10), S3C2440_GPD10_SPICLK1);
		s3c2410_gpio_cfgpin(S3C2410_GPD(9), S3C2440_GPD9_SPIMOSI1);
		s3c2410_gpio_cfgpin(S3C2410_GPD(8), S3C2440_GPD8_SPIMISO1);
-		s3c2410_gpio_pullup(S3C2410_GPD(10), 0);
-		s3c2410_gpio_pullup(S3C2410_GPD(9), 0);
+		s3c2410_gpio_pullup(S3C2410_GPD(10), 1);
+		s3c2410_gpio_pullup(S3C2410_GPD(9), 1);
	} else {
		s3c2410_gpio_cfgpin(S3C2410_GPD(8), S3C2410_GPIO_INPUT);
		s3c2410_gpio_cfgpin(S3C2410_GPD(9), S3C2410_GPIO_INPUT);
-		s3c2410_gpio_pullup(S3C2410_GPD(10), 1);
-		s3c2410_gpio_pullup(S3C2410_GPD(9), 1);
-		s3c2410_gpio_pullup(S3C2410_GPD(8), 1);
+		s3c2410_gpio_pullup(S3C2410_GPD(10), 0);
+		s3c2410_gpio_pullup(S3C2410_GPD(9), 0);
+		s3c2410_gpio_pullup(S3C2410_GPD(8), 0);
	}
 }
--- linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus1-gpg5_6_7.c	2010-05-04 03:28:43.000000000 -0300
+++ linux-2.6.32/arch/arm/plat-s3c24xx/spi-bus1-gpg5_6_7.c	2010-07-02 14:28:18.000000000 -0300
@@ -24,13 +24,13 @@
		s3c2410_gpio_cfgpin(S3C2410_GPG(7), S3C2410_GPG7_SPICLK1);
		s3c2410_gpio_cfgpin(S3C2410_GPG(6), S3C2410_GPG6_SPIMOSI1);
		s3c2410_gpio_cfgpin(S3C2410_GPG(5), S3C2410_GPG5_SPIMISO1);
-		s3c2410_gpio_pullup(S3C2410_GPG(5), 0);
-		s3c2410_gpio_pullup(S3C2410_GPG(6), 0);
+		s3c2410_gpio_pullup(S3C2410_GPG(5), 1);
+		s3c2410_gpio_pullup(S3C2410_GPG(6), 1);
	} else {
		s3c2410_gpio_cfgpin(S3C2410_GPG(7), S3C2410_GPIO_INPUT);
		s3c2410_gpio_cfgpin(S3C2410_GPG(5), S3C2410_GPIO_INPUT);
-		s3c2410_gpio_pullup(S3C2410_GPG(5), 1);
-		s3c2410_gpio_pullup(S3C2410_GPG(6), 1);
-		s3c2410_gpio_pullup(S3C2410_GPG(7), 1);
+		s3c2410_gpio_pullup(S3C2410_GPG(5), 0);
+		s3c2410_gpio_pullup(S3C2410_GPG(6), 0);
+		s3c2410_gpio_pullup(S3C2410_GPG(7), 0);
	}
 }
--- linux-2.6.32/arch/arm/mach-s3c2440/mach-mini2440.c	2010-05-04 03:28:43.000000000 -0300
+++ linux-2.6.32/arch/arm/mach-s3c2440/mach-mini2440.c	2010-07-02 14:37:55.000000000 -0300
@@ -761,7 +761,7 @@
	s3c2410_gpio_cfgpin(S3C2410_GPG(4), S3C2410_GPIO_OUTPUT);

	/* remove pullup on optional PWM backlight -- unused on 3.5 and 7"s */
-	s3c2410_gpio_pullup(S3C2410_GPB(1), 0);
+	s3c2410_gpio_pullup(S3C2410_GPB(1), 1);
	s3c2410_gpio_setpin(S3C2410_GPB(1), 0);
	s3c2410_gpio_cfgpin(S3C2410_GPB(1), S3C2410_GPIO_INPUT);

@@ -770,7 +770,7 @@

	/* mark the key as input, without pullups (there is one on the board) */
	for (i = 0; i < ARRAY_SIZE(mini2440_buttons); i++) {
-		s3c2410_gpio_pullup(mini2440_buttons[i].gpio, 0);
+		s3c2410_gpio_pullup(mini2440_buttons[i].gpio, 1);
		s3c2410_gpio_cfgpin(mini2440_buttons[i].gpio,
					S3C2410_GPIO_INPUT);
	}
